### PR TITLE
ENH: stats: Implement _munp() for the skewnorm distribution.

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
The skew-normal distribution has a moment-generating function from
which we can derive the noncentral moments.  This allows _munp(order, a),
the method of the rv_continuous class that computes the "standard" (i.e.
loc=0, scale=1) noncentral moment, to be implemented.  This is much
faster and more accurate than the default method, which uses
scipy.integrate.quad to compute the moments.

### CODE REVIEW

**Comment by larsoner** [comment]
#16487 has been merged, if you rebase or merge with `upstream` you should be able to remove the `pytest.mark.timeout` mark here:

https://github.com/scipy/scipy/blob/fb0d2d2c8130963b7e49c580a8e261162f57f137/scipy/stats/tests/test_continuous_basic.py#L269

**Comment by WarrenWeckesser** [comment]
@larsoner wrote
> you should be able to remove the pytest.mark.timeout mark

Done.  I left the code for handling custom marks in `test_moments`, with a comment about how it might be used again in the future.  It could all be removed instead, now that it is not being used; I'm fine with that option too.

**Comment by larsoner** [comment]
Either way is fine by me. It'll be great just to see this come back green, it'll be a nice speedup!

File: scipy/stats/_continuous_distns.py
Comment (larsoner) [inline scipy/stats/_continuous_distns.py:7671]: This will slow down the import of `_continuous_distns.py` itself, and thus presumably `scipy.stats`. Instead this could be put in a `@lru_cache`-decorated `_get_skewnorm_odd_moments()` function that returns these values. Then there is no cost at import time, only at the first skewnorm call.

File: scipy/stats/_precompute/skewnorm_generate_noncentral_moments.py
Comment (larsoner) [inline scipy/stats/_precompute/skewnorm_generate_noncentral_moments.py:5]: Thanks for the explanation, I was confused at first...

File: scipy/stats/tests/test_distributions.py
Comment (larsoner) [inline scipy/stats/tests/test_distributions.py:2722]: Same here, this will incur a cost at the test collection phase, even if you are doing `pytest scipy -k test_firwin` or so. You could instead make this a `pytest.fixture(scope='module')` that returns these values, then use it inside the function. Then the cost is only incurred during setup of the specific test(s) that need these values.

**Comment by WarrenWeckesser** [comment]
> It'll be great just to see this come back green...

I don't expect this latest commit to come back all green, because of a recurring unrelated failure in sparse.linalg that is happening in the `Main prerelease_deps_coverage_64bit_blas` job.

File: scipy/stats/_continuous_distns.py
Comment (WarrenWeckesser) [reply scipy/stats/_continuous_distns.py:7671]: The joys of micro-optimizing Python...  I'll give it a shot.

It would be interesting to see how much the import time of any specific module could be improved if we went over the entire code base with an eye towards this type of optimization.

Comment (larsoner) [reply scipy/stats/_continuous_distns.py:7671]: Maybe it's not worth it if it's in the microsecond range. But in the past at least I've nested `scipy` submodule imports in functions because they were slow. Hence the idea to at least work toward not making the problem worse...

File: scipy/stats/_precompute/skewnorm_generate_noncentral_moments.py
Comment (WarrenWeckesser) [reply scipy/stats/_precompute/skewnorm_generate_noncentral_moments.py:5]: I followed @person142's example from `special` and tucked this away in the new `_precompute` directory.  The idea is to stash here the one-off scripts (typically based on sympy or mpmath) that we use to generate other bits of code.

File: scipy/stats/tests/test_distributions.py
Comment (WarrenWeckesser) [reply scipy/stats/tests/test_distributions.py:2722]: Can a fixture be used as an argument to `pytest.mark.parametrize()`?

Comment (WarrenWeckesser) [reply scipy/stats/tests/test_distributions.py:2722]: Apparently that's an old FAQ: https://github.com/pytest-dev/pytest/issues/349

File: scipy/stats/_continuous_distns.py
Comment (WarrenWeckesser) [reply scipy/stats/_continuous_distns.py]: Done.

File: scipy/stats/tests/test_distributions.py
Comment (WarrenWeckesser) [reply scipy/stats/tests/test_distributions.py:2722]: @larsoner, I'm punting on the fixture stuff.  If you know exactly what you want to do, perhaps you could do the optimization in a follow-up PR, or just push your change to this PR?

**Comment by WarrenWeckesser** [comment]
The `Main prerelease_deps_coverage_64bit_blas` job is passing now (thanks @rgommers!), but now the Azure job `Main source_distribution` is failing in all the PRs.

**Comment by rgommers** [comment]
> but now the Azure job `Main source_distribution` is failing in all the PRs.

already fixed by gh-16541

**Review by mdhaber — APPROVED** [review]
Approved from an accuracy standpoint. I wrote code independently to check the coefficients, walked through how the coefficients turn into moments, manually checked unit test values against Wolfram Alpha, tested `skewnorm.moment` on the command line, made sure unit test was testing what it's purported to, etc. 

@tupui would you mind just looking over style? There are things here that seem to be done for maximum speed that make review more difficult, like the double factorial function being written out and being expressed differently than in `special.factorial2`, but I thought I'd ask for a second opinion before I comment on the rest. If it looks fine to you, I'll go ahead and merge, but I thought someone else should comment on the subjective stuff before I do.

File: scipy/stats/_continuous_distns.py
Comment (tupui) [inline scipy/stats/_continuous_distns.py:7674]: This could be in the class and not use lru but a simple `if` to check if an attribute is filled or not.

Only reason I see to go that route is if we think multiple classes would be created by user and we would want to keep things around in memory.

**Review by tupui — COMMENTED** [review]
I just have 2 suggestions. The rest LGTM, the factorial part has enough comment so I am fine with that.

File: scipy/stats/_continuous_distns.py
Comment (WarrenWeckesser) [reply scipy/stats/_continuous_distns.py]: I changed the function to a method of the class.  Initially I kept the use of @lru_cache, but that is overkill--there is only one value, and it is cached the first time it is used.  An `if` could be used, but that is really just an implementation detail for a cached property.  In Python 3.8, there is [`functools.cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property), which does exactly what we want: the dictionary is constructed the first time it is referenced and becomes an attribute of the instance, so we avoid the cost of creation of the dictionary during import, and subsequent references use the attribute, so there is not even the cost of an `if` statement.

Comment (tupui) [reply scipy/stats/_continuous_distns.py]: Fine by me

File: scipy/stats/_precompute/skewnorm_generate_noncentral_moments.py
Comment (WarrenWeckesser) [reply scipy/stats/_precompute/skewnorm_generate_noncentral_moments.py:5]: The repo is the *best* place to keep this.   It is not noise, it is a code generation tool.   Future devs (including our future selves) will thank us for keeping this code around. More than once, I've wished I had the code that was used to generate the various polynomials and rational functions that are used throughout `scipy.special`.  Having it here will allow future devs to verify the coefficients, and to possibly use variations of the script to generate code for other distributions.

I had used the Unicode character δ in the script because it made the output that is printed to the terminal look nice, but it is not important, so I switched it to `d`.  (There is actually no reason we couldn't allow δ in the Unicode check; we already have a few of the other Greek letters that are commonly used in mathematics.  But that can wait until someone finds a compelling reason to use it in a public docstring.)

**Comment by WarrenWeckesser** [comment]
@tupui, thanks for the review!  I have responded to your suggestions, and after merging `main` to resolve conflicts, I push some updates that address your comments.

If/when this gets merged, the commits should be squashed into one commit.  The commit message from the first commit (header line and body) are still probably all that is needed in the squashed commit message.

**Review by tupui — COMMENTED** [review]
> @tupui, thanks for the review! I have responded to your suggestions, and after merging main to resolve conflicts, I pushed some updates that address your comments.

Thank you, I still have my concern around the added file. See the discussion.

> If/when this gets merged, the commits should be squashed into one commit. The commit message from the first commit (header line and body) are still probably all that is needed in the squashed commit message.

Sure thank you.

**Comment by ev-br** [comment]
FWIW and from 30 thousand feet, I'd welcome this sorts of scripts. They are a big help in "what are these magic coefficients and where did they come from" situations. Where to keep them, a folder named _precompute looks reasonable to me, it's a single `ls-l` away. A different easily discoverable place would work, too, though. Not sure what that would be?

File: scipy/stats/_precompute/skewnorm_generate_noncentral_moments.py
Comment (WarrenWeckesser) [reply scipy/stats/_precompute/skewnorm_generate_noncentral_moments.py]: OK, I removed the script.   I'm fine with storing it somewhere else.  We can create new top-level repo for code generation and analysis scripts, useful docs, and anything else that we use for SciPy development that is worth saving but that doesn't quite fit into the structure of the main repo.  Who has the appropriate permissions for creating such a repo?

Comment (tupui) [reply scipy/stats/_precompute/skewnorm_generate_noncentral_moments.py]: Thank you Warren. All maintainers should be able to create a repo. Seems like I can, so I am sure you can too. We currently have archive. Maybe that could go there? I am fine with having a new one as well. Maybe this calls for an email on the list.

**Comment by mdhaber** [comment]
> But the code itself is not good for all that. It just make the boat heavier.... @mdhaber do you have an opinion on this?

Yes, this is one of the things I hoped you would comment on.  I confirmed these coefficients by writing my own code to generate them independently, so I don't think should also have to review line-by-line the code used by the author to generate them - especially if it's a script that itself writes Python code. I think it's is valuable code, but I don't think it belongs here in SciPy proper.

The `lru_cache` was another thing that concerned me, and so was the use of `δ` in the code. Both of those were removed, so that's good. I'm still not a fan of rewriting and - without comment - refactoring `factorial2`, but it seems to be correct, so I'll go ahead and merge.

**Pull request merged**
